app/src/forms/submission/service.js

```diff
@@ -0,0 +1,100 @@
+const { Form, FormVersion, FormSubmission, FormSubmissionUserPermissions, SubmissionMetadata, UserFormAccess } = require('../common/models');
+
+const Permissions = require('../common/constants').Permissions;
+
+const Problem = require('api-problem');
+const {transaction} = require('objection');
+
+const service = {
+
+  _fetchSubmissionData: async(formSubmissionId) => {
+    const meta = await SubmissionMetadata.query()
+      .where('submissionId', formSubmissionId)
+      .first()
+      .throwIfNotFound();
+
+    const submission = await FormSubmission.query()
+      .findById(meta.submissionId)
+      .throwIfNotFound();
+    const version = await FormVersion.query()
+      .findById(meta.formVersionId)
+      .throwIfNotFound();
+    const form = await Form.query()
+      .findById(meta.formId)
+      .allowGraph('identityProviders')
+      .withGraphFetched('identityProviders(orderDefault)')
+      .throwIfNotFound();
+
+    return {
+      form: form,
+      version: version,
+      submission: submission
+    };
+  },
+
+  read: async (formSubmissionId, currentUser, permissions = [Permissions.SUBMISSION_READ]) => {
+    const result = await service._fetchSubmissionData(formSubmissionId);
+
+    const checkFormSubmissionsPermission = async () => {
+      if (currentUser.public) return false;
+      return UserFormAccess.query()
+        .modify('filterFormId', result.form.id)
+        .modify('filterUserId', currentUser.id)
+        .modify('filterByAccess', null, null, permissions)
+        .first();
+    };
+
+    const checkSubmissionPermission = async () => {
+      if (currentUser.public) return false;
+      return FormSubmissionUserPermissions.query()
+        .modify('filterSubmissionId', formSubmissionId)
+        .modify('filterUserId', currentUser.id)
+        .modify('filterByPermissions', permissions)
+        .first();
+    };
+
+    const isDeleted = result.submission.deleted;
+    const isDraft = result.submission.draft;
+    const publicAllowed = result.form.identityProviders.find(p => p.code === 'public');
+    const idpAllowed = result.form.identityProviders.find(p => p.code === currentUser.idp);
+    const formSubmissionsPermission = await checkFormSubmissionsPermission();
+    const submissionPermission = await checkSubmissionPermission();
```

Comment:
Depending on speed of this, would the second lookup potentially be avoidable based on the formSubmissionsPermission?

Reply:
logically, there would be a path to optimize this, but until we need to do so, i'd rather leave it.  it will be a lot easier to debug with a straightforward logic path.  if speed is an issue, we can shuffle this around.

Reply:
changed it up, calls when next permission set needed.